presto-native-execution/presto_cpp/main/runtime-metrics/PrometheusStatsReporter.cpp

```diff
@@ -0,0 +1,238 @@
+
+namespace facebook::presto::prometheus {
+
+// Initialize singleton for the reporter
+folly::Singleton<facebook::velox::BaseStatsReporter> reporter(
+    []() -> facebook::velox::BaseStatsReporter* {
+      return facebook::presto::prometheus::PrometheusStatsReporter::
+          createPrometheusReporter()
+              .release();
+    });
+
+static constexpr std::string_view kSummarySuffix("_summary");
+
+struct PrometheusStatsReporter::PrometheusImpl {
+  explicit PrometheusImpl(const ::prometheus::Labels& labels) {
+    registry = std::make_shared<::prometheus::Registry>();
+    for (const auto& itr : labels) {
+      this->labels[itr.first] = itr.second;
+    }
+  }
+
+  ::prometheus::Labels labels;
+  std::shared_ptr<::prometheus::Registry> registry;
+};
+
+PrometheusStatsReporter::PrometheusStatsReporter(
+    const std::map<std::string, std::string>& labels) {
+  impl_ = std::make_shared<PrometheusImpl>(labels);
+}
+
+void PrometheusStatsReporter::registerMetricExportType(
+    const char* key,
+    facebook::velox::StatType statType) const {
+  if (registeredMetricsMap_.count(key)) {
+    VLOG(2) << "Trying to register already registered metric " << key;
+    return;
+  }
+  // '.' is replaced with '_'.
+  std::string sanitizedMetricKey = std::string(key);
+  std::replace(sanitizedMetricKey.begin(), sanitizedMetricKey.end(), '.', '_');
+  switch (statType) {
+    case facebook::velox::StatType::COUNT: {
+      // A new MetricFamily object is built for every new metric key.
+      auto& counterFamily = ::prometheus::BuildCounter()
+                                .Name(sanitizedMetricKey)
+                                .Register(*impl_->registry);
+      auto& counter = counterFamily.Add(impl_->labels);
+      registeredMetricsMap_.emplace(
+          std::string(key), StatsInfo{statType, &counter});
+    } break;
+    case facebook::velox::StatType::SUM:
+    case facebook::velox::StatType::AVG:
+    case facebook::velox::StatType::RATE: {
+      auto& gaugeFamily = ::prometheus::BuildGauge()
+                              .Name(sanitizedMetricKey)
+                              .Register(*impl_->registry);
+      auto& gauge = gaugeFamily.Add(impl_->labels);
+      registeredMetricsMap_.emplace(
+          std::string(key), StatsInfo{statType, &gauge});
+    } break;
+    default:
+      VELOX_UNSUPPORTED(
+          "Unsupported metric type {}", velox::statTypeString(statType));
+  }
+}
+
+void PrometheusStatsReporter::registerMetricExportType(
+    folly::StringPiece key,
+    facebook::velox::StatType statType) const {
+  registerMetricExportType(key.toString().c_str(), statType);
+}
+
+void PrometheusStatsReporter::registerHistogramMetricExportType(
+    const char* key,
+    int64_t bucketWidth,
+    int64_t min,
+    int64_t max,
+    const std::vector<int32_t>& pcts) const {
+  if (registeredMetricsMap_.count(key)) {
+    // Already registered;
+    VLOG(1) << "Trying to register already registered metric " << key;
```

Comment:
VLOG(2)

Reply:
I feel these logs must be VLOG(1). We need to know at least from logs if something went wrong. Also, they are not too verbose.